src/mpid/ch4/src/ch4_part.c

```diff
@@ -0,0 +1,128 @@
+    MPIR_ERR_CHKANDSTMT((req) == NULL, mpi_errno, MPIX_ERR_NOREQ, goto fn_fail, "**nomemreq");
+
+    MPIR_Comm_add_ref(comm);
+    req->comm = comm;
+
+    MPIR_Datatype_add_ref_if_not_builtin(datatype);
+    MPIDI_PART_REQUEST(req, datatype) = datatype;
+
+    MPIDI_PART_REQUEST(req, rank) = rank;
+    MPIDI_PART_REQUEST(req, tag) = tag;
+    MPIDI_PART_REQUEST(req, context_id) = comm->context_id;
+    MPIDI_PART_REQUEST(req, buffer) = buf;
+    MPIDI_PART_REQUEST(req, count) = count;
+    MPIDI_PART_REQUEST(req, active_flag) = 0;
+    req->u.part.partitions = partitions;
+
+    /* Inactive partitioned comm request can be freed by request_free.
+     * Completion cntr increases when request becomes active at start. */
+    MPIR_cc_set(req->cc_ptr, 0);
+
+  fn_exit:
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPID_PART_REQ_CREATE);
+    return req;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPID_Psend_init(void *buf, int partitions, MPI_Count count,
+                    MPI_Datatype datatype, int dest, int tag,
+                    MPIR_Comm * comm, MPIR_Info * info, MPIR_Request ** request)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPID_PSEND_INIT);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPID_PSEND_INIT);
+
+    MPIDI_av_entry_t *av = MPIDIU_comm_rank_to_av(comm, dest);
+
+    MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(0).lock);
+    MPIDI_workq_vci_progress_unsafe();
+
+    *request = part_req_create((void *) buf, partitions, count, datatype, dest, tag, comm,
+                               MPIR_REQUEST_KIND__PART_SEND);
+#ifdef MPIDI_CH4_DIRECT_NETMOD
+    mpi_errno = MPIDI_NM_mpi_psend_init(buf, partitions, count, datatype, dest, tag, comm,
+                                        info, av, request);
+#else
+    int is_local = MPIDI_av_is_local(av);
+    MPIDI_REQUEST(*request, is_local) = is_local;       /* use at start, pready, parrived */
+
+    if (is_local) {
+        mpi_errno = MPIDI_SHM_mpi_psend_init(buf, partitions, count, datatype, dest, tag,
+                                             comm, info, av, request);
+    } else {
+        mpi_errno = MPIDI_NM_mpi_psend_init(buf, partitions, count, datatype, dest, tag,
+                                            comm, info, av, request);
+    }
+#endif
+    MPIR_ERR_CHECK(mpi_errno);
+
+  fn_exit:
+    MPID_THREAD_CS_EXIT(VCI, MPIDI_VCI(0).lock);
+    MPIR_FUNC_VERBOSE_EXIT(MPID_STATE_MPID_PSEND_INIT);
+    return mpi_errno;
+  fn_fail:
+    goto fn_exit;
+}
+
+int MPID_Precv_init(void *buf, int partitions, MPI_Count count,
+                    MPI_Datatype datatype, int source, int tag,
+                    MPIR_Comm * comm, MPIR_Info * info, MPIR_Request ** request)
+{
+    int mpi_errno = MPI_SUCCESS;
+
+    MPIR_FUNC_VERBOSE_STATE_DECL(MPID_STATE_MPID_PRECV_INIT);
+    MPIR_FUNC_VERBOSE_ENTER(MPID_STATE_MPID_PRECV_INIT);
+
+    MPIDI_av_entry_t *av = MPIDIU_comm_rank_to_av(comm, source);
+
+    MPID_THREAD_CS_ENTER(VCI, MPIDI_VCI(0).lock);
```

Comment:
Could you take the vci lock inside device layer as the pt2pt path does. I have #4746 also moving the vci lock down in the RMA path. If you do it here, someone have to come here later to move them down again for extra work.

Reply:
I actually don't follow the vci design well. Will make the change as you suggested.

Reply:
Btw, I am not sure when we need call `MPIDI_workq_vci_progress` together with the vci lock. Based on #4746 , is it needed only at `MPID` layer when `MPIDI_CH4_USE_WORK_QUEUES` is on?

Reply:
One more question regarding vci lock for `psend_init|precv_init`. As I think they are more similar to `win_create` rather than regular pt2pt, I am trying to find where we do lock for `win_create`. But not found in neither `MPID` nor `MPIDIG`. Do we lock such resource allocation calls at MPI layer?

Reply:
Right. Then we still need lock for `psend|precv_init` as two threads may call it on the same comm. Thanks for explanation.